File: spring-web-java/src/main/java/spring/web/java/infrastructure/config/security/JwtTokenProvider.java
package spring.web.java.infrastructure.config.security;

import java.time.Duration;
import java.util.Date;
import java.util.List;

import javax.crypto.SecretKey;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;

@Component
public class JwtTokenProvider {

	private final SecretKey accessTokenKey;
	private final long accessTokenExpirationTime;
	private final SecretKey refreshTokenKey;
	private final long refreshTokenExpirationTime;

	public JwtTokenProvider(
		@Value("${jwt.access-token.key}") String accessTokenKey,
		@Value("${jwt.access-token.expiration-duration}") Duration accessTokenExpirationDuration,
		@Value("${jwt.refresh-token.key}") String refreshTokenKey,
		@Value("${jwt.refresh-token.expiration-duration}") Duration refreshTokenExpirationDuration
	) {
		this.accessTokenKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(accessTokenKey));
		this.accessTokenExpirationTime = accessTokenExpirationDuration.toMillis();
		this.refreshTokenKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(refreshTokenKey));
		this.refreshTokenExpirationTime = refreshTokenExpirationDuration.toMillis();
	}

	public String createAccessToken(Long memberId, List<String> permissions) {
		return Jwts.builder()
			.claim("memberId", memberId)
			.claim("permissions", permissions)
			.issuedAt(new Date())
			.expiration(new Date(new Date().getTime() + accessTokenExpirationTime))
			.signWith(accessTokenKey)
			.compact();
	}

	public String createRefreshToken() {
		return Jwts.builder()
			.issuedAt(new Date())
			.expiration(new Date(new Date().getTime() + refreshTokenExpirationTime))
			.signWith(refreshTokenKey)
			.compact();
	}

	public Claims parseAccessToken(String token) {
		return Jwts.parser()
			.verifyWith(accessTokenKey)
			.build()
			.parseSignedClaims(token)
			.getPayload();
	}

	public void validateRefreshToken(String token) {
		Jwts.parser()
			.verifyWith(refreshTokenKey)
			.build()
			.parseSignedClaims(token);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/infrastructure/config/security/JwtTokenProvider.kt
package spring.web.kotlin.infrastructure.config.security

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.time.Duration
import java.util.*

@Component
class JwtTokenProvider(
    @Value("\${jwt.access-token.key}") accessTokenKey: String,
    @Value("\${jwt.access-token.expiration-duration}") accessTokenExpirationDuration: Duration,
    @Value("\${jwt.refresh-token.key}") refreshTokenKey: String,
    @Value("\${jwt.refresh-token.expiration-duration}") refreshTokenExpirationDuration: Duration,
) {
    private val accessTokenKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(accessTokenKey))
    private val accessTokenExpirationTime = accessTokenExpirationDuration.toMillis()
    private val refreshTokenKey = Keys.hmacShaKeyFor(Decoders.BASE64.decode(refreshTokenKey))
    private val refreshTokenExpirationTime = refreshTokenExpirationDuration.toMillis()

    fun createAccessToken(memberId: Long, permissions: List<String>): String =
        Jwts.builder()
            .claim("memberId", memberId)
            .claim("permissions", permissions)
            .issuedAt(Date())
            .expiration(Date(Date().time + accessTokenExpirationTime))
            .signWith(accessTokenKey)
            .compact()

    fun createRefreshToken(): String =
        Jwts.builder()
            .issuedAt(Date())
            .expiration(Date(Date().time + refreshTokenExpirationTime))
            .signWith(refreshTokenKey)
            .compact()

    fun parseAccessToken(token: String): Claims =
        Jwts.parser()
            .verifyWith(accessTokenKey)
            .build()
            .parseSignedClaims(token)
            .payload

    fun parseRefreshToken(token: String): Claims =
        Jwts.parser()
            .verifyWith(refreshTokenKey)
            .build()
            .parseSignedClaims(token)
            .payload
}